Add HPC docs page. (#278)
Cambridge-ICCS/FTorch: pages/hpc.md

title: Guidance for use in High Performance Computing (HPC)

[TOC]

A common application of FTorch (indeed, the driving one for development) is the
coupling of machine learning components to models running on HPC systems.

> Here we provide some guidance/hints to help with deploymwent in these settings.

Reviewer: ```suggestion
Here we provide some guidance/hints to help with deployment in these settings.
```

## Installation

### Building for basic use

The basic installation procedure is the same as described in the

> [main documentation](page//cmake.html) and README, cloning from

Reviewer: Is this link right? Should it not be pages?

Author: Yes, thank you. Fixed in https://github.com/Cambridge-ICCS/FTorch/pull/278/commits/06130e49df1c23d9b1520be6456fa9596e7ff0cc

[GitHub](https://github.com/Cambridge-ICCS/FTorch) and building using CMake.

### Obtaining LibTorch

For use on a HPC system we advise linking to an installation of LibTorch rather than

> installing full PyTorch.

Reviewer: Perhaps comment on the fact that the development of the PyTorch ML model would be done off the HPC?

Author: How do you feel about https://github.com/Cambridge-ICCS/FTorch/pull/278/commits/06130e49df1c23d9b1520be6456fa9596e7ff0cc

Suggested 'Python/PyTorch development is done elsewhere' as it may be that the HPC is also used to train if distributed/accelerated.

This will reduce the dependencies and remove any requirement of Python.
LibTorch can be obtained from the
[PyTorch website](https://pytorch.org/get-started/locally/).

Once you have successfully tested and deployed FTorch in your code we recommend speaking
to your administrator/software stack manager to make your chosen version of libtorch
loadable as a `module`.
This will improve reproducibility and simplify the process for future users on your
system.
See the [information below](#libtorch-as-a-module) for further details.

### Environment management

It is important that FTorch is built using the same environment and compilers as the
software to which it will be linked.

> Therefore before running the build you should ensure that you match the environment to

Reviewer: ```suggestion
Therefore before starting the build you should ensure that you match the environment to
```

that which your code will be built with.
This will usually be done by using the same `module` commands as you would use to build
the model:
```sh
module purge
module load ...
```

Alternatively you may be provided with a shell script that runs these commands and sets
environment variables etc. that can be sourced:
```sh
source model_environment.sh
```

Complex models with custom build systems may obfuscate this process, and you might need
to probe the build system/scripts for this information.
If in doubt speak to the maintainer of the software for your system, or the manager of
the software stack on the machine.

> Because of the need to match compilers it is strongly recommended to use the

Reviewer: ```suggestion
Because of the need to match compilers it is strongly recommended to specify the
```

```sh
-L<path/to/FTorch/install/location>/lib64 -lftorch
```
This is often done by appending to an `LDFLAGS` linker flags variable or similar:
```sh
LDFLAGS += -L<path/to/FTorch/install/location>/lib64 -lftorch
```

You may also need to add the location of the dynamic library `.so` files to your
`LD_LIBRARY_PATH` environment variable unless installing in a default location:
```sh
export LD_LIBRARY_PATH = $LD_LIBRARY_PATH:<path/to/FTorch/installation>/lib64
```

> Note: _Depending on your system and architecture `lib` may be `lib64` or something similar._

> Note: _On MacOS devices you will need to set `DYLD_LIBRARY_PATH` rather than `LD_LIBRARY_PATH`._

Whilst experimenting it may be useful to build FTorch using the `CMAKE_BUILD_TYPE=RELEASE`
CMake flag to allow useful error messages and investigation with debugging tools.

### Module systems

Most HPC systems are managed using [Environment Modules](https://modules.sourceforge.net/).
To build FTorch it is important you
[match the environment in which you build FTorch to that of the executable](#environment-management)
by loading the same modules as when building the main code.

> As a minimal requirement you will need to load modules for compilers and cmake.

Reviewer: ```suggestion
As a minimal requirement you will need to load modules for compilers and CMake.
```

Further functionalities may require loading of additional modules such as an
MPI installation and CUDA.

> Some systems may also have pFUnit available ada loadable module to save you needing to

Reviewer: ```suggestion
Some systems may also have pFUnit available as a loadable module to save you needing to
```

> build from scratch per the documentation.

Reviewer: Perhaps prefix with "In the case that you want to build FTorch's test suite"?

Author: Appended in https://github.com/Cambridge-ICCS/FTorch/pull/278/commits/06130e49df1c23d9b1520be6456fa9596e7ff0cc

#### LibTorch as a module

Once you have a working build of FTorch it is advisable to pin the version of LibTorch
and make it a loadable module to improve reproducibility and simplify the build process
for subsequent users on the system.

> This can be done my the software manager after which you can use

Reviewer: ```suggestion
This can be done by the software manager after which you can use
```

```sh
module load libtorch
```
or similar instead of downloading the binary from the PyTorch website.

Note that the module name on your system may include additional information about the
version, compilers used, and a hash code.

#### FTorch as a module 

If there are many users who want to use FTorch on a system it may be worth building
and making it loadable as a module itself.
The module should be labelled with the compilers it was built with (see the

> [importance of enviroinment matching](#environment-management)) and automatically load

Reviewer: ```suggestion
[importance of environment matching](#environment-management)) and automatically load
```
